lib/matrix_client/poller.ex

```diff
@@ -75,18 +85,29 @@ defmodule M51.MatrixClient.Poller do
     # Need to be larger than the timeout above (both in milliseconds)
     options = [recv_timeout: 1_000_000]
 
+    delay =
+      if delay do
+        Process.sleep(delay)
+        Kernel.min(delay * @connect_delay_factor, @connect_delay_max)
+      else
+        @connect_delay_min
+      end
+
     case M51.Matrix.RawClient.get(raw_client, path, [], options) do
       {:ok, events} ->
         handle_events(sup_pid, is_backlog, events)
         events["next_batch"]
 
       {:error, code, _} when code >= 500 and code < 600 ->
-        # server error, try again
-        poll_one(sup_pid, since, raw_client)
+        # server request processing error, try again
+        poll_one(sup_pid, since, raw_client, delay)
 
-      {:error, nil, :closed} ->
-        # server closed connection, likely due to timeout, retry
-        poll_one(sup_pid, since, raw_client)
+      {:error, nil, reason} ->
+        if delay do
+          Logger.warn("Server reconnection error [#{reason}], will retry in #{delay}s")
```

Comment:
Actually it will retry in `Kernel.min(delay * @connect_delay_factor, @connect_delay_max)`, not in `delay`. (And should be divided by 1000 to get seconds.)

Reply:
Um, no? delay is adjusted AFTER sleep, so this will be the delay.
True about units, will fix as mentioned.